### ISSUE TITLE
Static reference to container makes integration testing sequential

### ISSUE THREAD

**Issue by smithybc** [autofac/Autofac.AspNetCore.Multitenant#5]
Hi

Due to the extension method requiring a reference to the container and that being static, it means that integration tests can't run in parallel when you want to change different dependencies for tests because the static reference is shared by the process. Even if you have a new WebHostBuilder for each test this reference will obviously be shared. Would it possible to expose another way of adding in this functionality without the static reference?

**Comment by tillig** [comment]
I'm looking through the code in the actual Autofac.Integration.AspNetCore.Multitenant assembly and the only static references I'm finding are to reflection related items. Static like `static`, that is. The keyword.

We do need to be able to get to the container, so the extension method requires a fixed reference to a function that will return the multitenant container. It doesn't actually execute that function; instead it's passed along and stored in the middleware so it can be used. How you implement that function is up to you. If you make it access a static property, yeah, it'll be a static reference. If you do something else to facilitate your integration tests, well, that's up to you.

And, of course, we can't add the accessor function to the container itself because we have a chicken/egg problem at that point - we need access to the _container_ not a _tenant-specific scope_ at the request middleware point and we're setting up the whole notion of `RequestServices` so we can't ask the current executing context to resolve that for us. Basically, you have to pass that in somehow... and that's the function.

But, again, that could be any function. You could store the container somewhere else in your web host, not a static property. The Sandbox sample shows static because that's the easiest way to go, but the Sandbox is a simple example just to verify functionality. Your app could be more complex to support your needs, but that's app-specific.

Perhaps you could share some more specific indication (maybe in the form of pseudocode or a more detailed description) of how exactly you'd like to see things changed. I may not be understanding what you're seeing or asking for.

**Comment by smithybc** [comment]
Thanks for the reply. You're right in what you say. I didn't explain myself very well and you're correct in thinking I'm talking in regards to the example given. I guess what I'm saying is that, I can't see how you would be able to pass anything other than a static reference to this function when using a Startup file and this extension method but I'm probably missing something. It would be handy if there was another extension that you could use within configure services within the startup that achieves the same thing. The classes are internal so you can't do this yourself unless you want to dupe all the code.

**Comment by tillig** [comment]
For the sake of understanding, let's say you did dupe the code. What would you have to do to accomplish what you're looking for?

- We have to generate `RequestServices` (which is the whole problem we're solving here).
- We need the actual container, not a tenant lifetime scope in the middleware where that happens.
- We need a nice, convenient extension method to allow people to set all that up (i.e., the 80% case).

So let's say we were doing all this from scratch or we were going to do major refactoring. How would you change things in the code here to make this all happen?

Or is this something you could change in the application's startup code to enable your scenario?

I recognize the challenge you're facing, but I'm at a loss for how we could change what we're doing and still solve the problem this library solves - correctly setting up `RequestServices` on each request.

**Comment by smithybc** [comment]
Maybe I have it all wrong but the configureServices code within that WebHostBuilder extension method would just need to be able to be called within the startup file. So you could create a non static reference and pass it to it directly after you have built the multitenantContainer within your startup. You could even create an extension for this on IServiceCollection and wrap it within the existing WebHostBuilder extension.

So within startup:

 ```
public IServiceProvider ConfigureServices(IServiceCollection services)
 {
    //for brevity we have added all our services and have built our MultitentContainer but to illustrate lets just assume I have a reference
 
           var container = new MultitenantContainer();
           services.ConfigureMultitenantRequestServices(container);
           
           return new AutofacServiceProvider(container);
 }
```


The ConfigureMultitenantRequestServices extension would just have this existing code:

```
public static IServiceCollection ConfigureMultitenantRequestServices(this IServiceCollection services)
{
     var descriptor = new ServiceDescriptor(typeof(IStartupFilter), sp => new MultitenantRequestServicesStartupFilter(multitenantContainerAccessor), ServiceLifetime.Transient);
                services.Insert(0, descriptor);
                services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();

}
```

I might have the wrong end of the stick because I haven't dug into much of the code at all and maybe you need this registered before the container is built?

**Comment by tillig** [comment]
You should probably be aware that [they're going to be removing the ability to return your own `IServiceProvider` from `ConfigureServices`](https://github.com/aspnet/Hosting/pull/1580#issuecomment-437636691) so adding more tacked on to that probably isn't a good thing. [I did mention to them that this will mess with the way we have to manage things like multitenancy](https://github.com/aspnet/Hosting/pull/1580#issuecomment-438372845) and the workaround will be that [during `Configure` you can capture the service provider factory instance](https://github.com/aspnet/Hosting/pull/1580#issuecomment-438512344) and get the container from there.

I don't honestly know how it'll all work together at the moment. I've not had a chance to look at that. I should probably file an issue over on the Autofac.Extensions.DependencyInjection stuff to update the docs and examples in anticipation of this.

Anyway, point being, the solution we arrive at here will need to be something you can set in that world; and since there's a lot of interest in moving toward things hanging right off the web host builder, having that available as the primary usage will be key.

**Comment by tillig** [comment]
Oh, I forgot, [I already did add an issue to update for that](https://github.com/autofac/Autofac.Extensions.DependencyInjection/issues/36).

**Comment by smithybc** [comment]
Thanks for the heads up! I was just following the autofac docs, although it's weird from .Net because at the moment they convert the Startup to meet the IStartup interface which has the interface that returns the IServiceProvider (disclaimer, this is what I read somewhere). Either way, at the moment the configureservices on the WebApplicationBuilder just invoke delegates of configureServices sequentially and the internals in a startup for ConfigureServices are treated the same way.

**Comment by joaope** [comment]
Hi @tillig, glad you didn't close the issue as I think there's one still, that can/should be addressed and definitely in today's _ASP Core 2.x_ world.

**The problem:**

Starting from this totally valid scenario of testing the application using `TestServer`:

```c#
[Theory]
[InlineData("tenant1", "response1")]
[InlineData("tenant2", "response2")]
public async Task TestMultitenantApplication(string tenantId, string expectedResponse)
{
    var server = new TestServer(new WebHostBuilder()
        .UseStartup<Startup>()
        .UseAutofacMultitenantRequestServices(() => Startup.ApplicationContainer));
    var client = server.CreateClient();

    var response = await client.GetStringAsync($"endpoint?tenant={tenantId}");

    Assert.Equal("some response", response);
}
```
For each fact being tested we know that we have a new instance of everything (Host, Startup and Client) but the container itself which is static. This wouldn't be a big of an issue if the container was always the same. Unfortunately it becomes problematic if the logic of creating the container is itself dynamic. 

```c#
public class Startup
{
    public static MultitenantContainer ApplicationContainer;

    public IServiceProvider Configure(IServiceCollection baseServices)
    {
        if (ForSomeReasonIWantThisContainer)
        {
            ApplicationContainer = new MultitenantContainer(...);
        }
        else
        {
            ApplicationContainer = new MultitenantContainer(...);
        }

        return new AutofacServiceProvider(ApplicationContainer);
    }
}
```
So now we have multiple instances of the application sharing the same container, which from a test's perspective is wrong for obvious reasons. I get that this might be a problem only during unit testing phase but a problem nonetheless, specially because we rely heavily on these kind of tests and they become flaky very quickly.

**Potential fix:**

Because of that I started playing with it and created a PoC of what a solution might be, by making an [`IServiceCollection.AddAutofacMultitenantRequestServices()`](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-d3a114f80067968c57ca12e9e3e6df4bR50) extension method.

It's the exact same logic of the original extension method but this way we can move things [out of the `IWebHostBuilder`](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-c14d72931e65699984aca54615513aaeL14) for it to be [closer to where is actually needed - the `Startup.ConfigureServices()` itself](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-e6abaa3b11c3a6209fdacc77c57ca1efL22). We can now have a clean instance of `Startup` that doesn't leak it's container (pun intended!).

I've also added some [hosted application's tests scenarios](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-9d3b549526553a796df6e7ed4ee4884f) just to make my point across although they are not really relevant to the solution's implementation.

**3.x world**

ASP Core 3.0 is shifting things quite a bit around this area and with that in mind the solution might look counter-intuitive, but I see this as two unrelated problems. Not only because they are one major version away from each other but also because whatever is happening with Autofac.Multitenant (and this package) to support those changes, it doesn't seem to just require internal and/or documentation change. This to say that if this `IServiceCollection.AddAutofacMultitenantRequestServices()` needs to be changed it's because `IWebHostBuilder.UseAutofacMultitenantRequestServices()` also needed to be changed.

Don't think there's a _right_ solution for this that satisfies both worlds. And specially because you don't even know how the 3.0 world will be. It's quite an interesting problem actually as it looks like `Startup.ConfigureServices` (and the Startup itself?) is rendered useless from an Autofac multitenancy perspective as you now rely on the IServiceProvider coming from it. Talking about the chicken/egg problem...

Let me know what your thoughts are. Happy to create a PR out of my branch if you think it's worth a discussion over there.

Thanks!

**Comment by joaope** [comment]
Hi @tillig, glad you didn't close the issue as I think there's one still, that can/should be addressed and definitely in today's _ASP Core 2.x_ world.

**The problem:**

Starting from this totally valid scenario of testing the application using `TestServer`:

```c#
[Theory]
[InlineData("tenant1", "response1")]
[InlineData("tenant2", "response2")]
public async Task TestMultitenantApplication(string tenantId, string expectedResponse)
{
    var server = new TestServer(new WebHostBuilder()
        .UseStartup<Startup>()
        .UseAutofacMultitenantRequestServices(() => Startup.ApplicationContainer));
    var client = server.CreateClient();

    var response = await client.GetStringAsync($"endpoint?tenant={tenantId}");

    Assert.Equal("some response", response);
}
```
For each fact being tested we know that we have a new instance of everything (Host, Startup and Client) but the container itself which is static. This wouldn't be a big of an issue if the container was always the same. Unfortunately it becomes problematic if the logic of creating the container is itself dynamic. 

```c#
public class Startup
{
    public static MultitenantContainer ApplicationContainer;

    public IServiceProvider Configure(IServiceCollection baseServices)
    {
        if (ForSomeReasonIWantThisContainer)
        {
            ApplicationContainer = new MultitenantContainer(...);
        }
        else
        {
            ApplicationContainer = new MultitenantContainer(...);
        }

        return new AutofacServiceProvider(ApplicationContainer);
    }
}
```
So now we have multiple instances of the application sharing the same container, which from a test's perspective is wrong for obvious reasons. I get that this might be a problem only during unit testing phase but a problem nonetheless, specially because we rely heavily on these kind of tests and they become flaky very quickly.

**Potential fix:**

Because of that I started playing with it and created a PoC of what a solution might be, by making an [`IServiceCollection.AddAutofacMultitenantRequestServices()`](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-d3a114f80067968c57ca12e9e3e6df4bR50) extension method.

It's the exact same logic of the original extension method but this way we can move things [out of the `IWebHostBuilder`](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-c14d72931e65699984aca54615513aaeL14) for it to be [closer to where is actually needed - the `Startup.ConfigureServices()` itself](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-e6abaa3b11c3a6209fdacc77c57ca1efL22). We can now have a clean instance of `Startup` that doesn't leak it's container (pun intended!).

I've also added some [hosted application's tests scenarios](https://github.com/autofac/Autofac.AspNetCore.Multitenant/compare/develop...joaope:avoid-startup-static-property#diff-9d3b549526553a796df6e7ed4ee4884f) just to make my point across although they are not really relevant to the solution's implementation.

**3.x world**

ASP Core 3.0 is shifting things quite a bit around this area and with that in mind the solution might look counter-intuitive, but I see this as two unrelated problems. Not only because they are one major version away from each other but also because whatever is happening with Autofac.Multitenant (and this package) to support those changes, it doesn't seem to just require internal and/or documentation changes. This to say that if `IServiceCollection.AddAutofacMultitenantRequestServices()` needs to be changed it's because `IWebHostBuilder.UseAutofacMultitenantRequestServices()` also needed to be changed.

Don't think there's a _right_ solution for this that satisfies both worlds. And specially because you don't even know how the 3.0 world will be. It's quite an interesting problem actually as it looks like `Startup.ConfigureServices()` (and the Startup itself?) is rendered useless from an Autofac multitenancy perspective as you now rely on the IServiceProvider coming from it. Talking about the chicken/egg problem...

Let me know what your thoughts are. Happy to create a PR out of my branch if you think it's worth a discussion over there.

Thanks!
